File: OptimizeService/app/dto/optimization/engine/problem_request.py
from __future__ import annotations

from typing import Optional, Union, List
from uuid import UUID
from pydantic import BaseModel, Field


class VehicleData(BaseModel):
    """Thông số thùng xe cho Optimization Engine (mm hoặc m, đồng bộ theo đơn vị)."""
    inner_l: float
    inner_w: float
    inner_h: float
    max_payload_kg: float
    door_w: Optional[float] = None
    door_h: Optional[float] = None


class PackageData(BaseModel):
    """Thông số một kiện hàng gửi tới Optimization Engine."""
    id: Union[UUID, str]
    l: float
    w: float
    h: float
    weight: float
    allowed_rotations: List[Union[int, str]] = Field(default_factory=list)
    stop_index: int = 0
    max_stack_weight_kg: Optional[float] = None
    fragile: bool = False


class StopData(BaseModel):
    """Điểm giao hàng trong chuyến."""
    id: Union[UUID, str, int]
    sequence: int


class PinnedData(BaseModel):
    """Kiện hàng đã được cố định vị trí trước (nếu có)."""
    package_id: Union[UUID, str]
    pos_x: float
    pos_y: float
    pos_z: float
    rotation_type: Union[int, str] = 0


class ProblemRequest(BaseModel):
    """Payload gửi tới Optimization Engine để tính toán phương án xếp hàng."""
    vehicle: VehicleData
    packages: List[PackageData]
    stops: List[StopData] = Field(default_factory=list)
    pinned: List[PinnedData] = Field(default_factory=list)
    objective: str = "MAX_VOLUME_UTIL"  # "MAX_VOLUME_UTIL" | "AXLE_BALANCE" | "MIN_HEIGHT"
    time_limit_sec: int = 60
    seed: Optional[int] = None

File: src/main/java/fu/se184491/loadmaster_be/dto/optimization/engine/ProblemRequest.java
package fu.se184491.loadmaster_be.dto.optimization.engine;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.util.List;

/**
 * DTO sent from Spring Boot → FastAPI Optimization Engine.
 *
 * FastAPI endpoint: POST /api/v1/optimization/jobs
 *
 * All dimensions are in millimetres (mm).
 * All weights are in kilograms (kg).
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class ProblemRequest {

    private VehicleData vehicle;
    private List<PackageData> packages;
    private List<StopData> stops;
    private List<PinnedData> pinned;

    /** "MAX_VOLUME_UTIL" | "MIN_HEIGHT" — mapped from OptimizationObjective */
    private String objective;

    /** Engine computation time ceiling in seconds (10–600) */
    private Integer timeLimitSec;

    /** Optional random seed for reproducibility */
    private Long seed;

    // ── Nested DTOs ──────────────────────────────────────────────────────────

    @Data @Builder @NoArgsConstructor @AllArgsConstructor
    public static class VehicleData {
        /** Inner cargo compartment dimensions (mm) */
        private Integer innerLength;
        private Integer innerWidth;
        private Integer innerHeight;

        /** Maximum total payload (kg) */
        private BigDecimal maxPayloadKg;

        /** Door opening dimensions — constrains package entry rotation (mm) */
        private Integer doorWidth;
        private Integer doorHeight;
    }

    @Data @Builder @NoArgsConstructor @AllArgsConstructor
    public static class PackageData {
        /** Spring Boot CargoPackage.id — engine echoes this back in result */
        private Long id;

        /** Physical dimensions (mm) */
        private Integer length;
        private Integer width;
        private Integer height;

        /** Actual measured weight (kg) */
        private BigDecimal weightKg;

        /**
         * Bitmask of allowed rotation types (0–5).
         * Computed from PackageType.allowRotateX/Y/Z flags.
         */
        private List<Integer> allowedRotations;

        /** Maximum stackable weight on top of this package (kg) */
        private BigDecimal maxStackWeightKg;

        /** Whether this package is fragile (affects stacking rules) */
        private Boolean fragile;

        /** Delivery stop index (0-based) within the trip's stop sequence */
        private Integer stopIndex;
    }

    @Data @Builder @NoArgsConstructor @AllArgsConstructor
    public static class StopData {
        private Long id;
        private Integer sequence;
    }

    @Data @Builder @NoArgsConstructor @AllArgsConstructor
    public static class PinnedData {
        private Long packageId;
        private Integer posX;
        private Integer posY;
        private Integer posZ;
        /** Rotation type (0–5) already fixed for this package */
        private Integer rotationType;
    }
}
